### PR TITLE
Fix Local Smoke Tests by Allowing ScriptSessionCacheInit's mkdir to Race in Parallel Tests

### PR DESCRIPTION
**Pull request by nbauernfeind**
Turns out that `NoConsoleSessionModule` is used in the testing context and if we simply do not require the `ScriptSessionCacheInit` then it no longer fails. I've left a few null checks in tearDown that fix cascading failures when setUp throws an exception in a super class.

### CODE REVIEW

File: server/src/test/java/io/deephaven/server/runner/DeephavenApiServerTestBase.java [hunk 1]
```diff
@@ -145,7 +143,9 @@ public void setUp() throws Exception {
 
     @After
     public void tearDown() throws Exception {
-        scopeCloseable.close();
+        if (scopeCloseable != null) {
+            scopeCloseable.close();
+        }
```
Comment (rcaudy) [inline server/src/test/java/io/deephaven/server/runner/DeephavenApiServerTestBase.java:148]: Should we null out the `scopeCloseable` out of paranoia?

File: server/src/main/java/io/deephaven/server/console/NoConsoleSessionModule.java [hunk 2]
```diff
@@ -28,8 +28,7 @@ ScriptSession bindScriptSession(NoLanguageDeephavenSession noLanguageSession) {
     @Provides
     NoLanguageDeephavenSession bindNoLanguageSession(
             @Named(PeriodicUpdateGraph.DEFAULT_UPDATE_GRAPH_NAME) final UpdateGraph updateGraph,
-            final OperationInitializer operationInitializer,
-            final ScriptSessionCacheInit ignored) {
+            final OperationInitializer operationInitializer) {
```
Comment (rcaudy) [inline server/src/main/java/io/deephaven/server/console/NoConsoleSessionModule.java:31]: This doesn't seem like exactly the right change. The issue appears to be with parallel execution, where one test creates the directory before another gets around to it.

Comment (rcaudy) [reply server/src/main/java/io/deephaven/server/console/NoConsoleSessionModule.java:31]:
```
Index: engine/table/src/main/java/io/deephaven/engine/util/AbstractScriptSession.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/engine/table/src/main/java/io/deephaven/engine/util/AbstractScriptSession.java b/engine/table/src/main/java/io/deephaven/engine/util/AbstractScriptSession.java
--- a/engine/table/src/main/java/io/deephaven/engine/util/AbstractScriptSession.java	(revision 25e0cb1ff86be7d7814085f8e096e9db4661f4bd)
+++ b/engine/table/src/main/java/io/deephaven/engine/util/AbstractScriptSession.java	(date 1715205168607)
@@ -26,6 +26,7 @@
 
 import java.io.File;
 import java.io.IOException;
+import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.*;
 import java.util.function.Function;
@@ -41,26 +42,27 @@
 
     private static final Path CLASS_CACHE_LOCATION = CacheDir.get().resolve("script-session-classes");
 
-    protected static File newClassCacheLocation() {
+    protected static Path newClassCacheLocation() {
         // TODO(deephaven-core#1713): Introduce instance-id concept
         final UUID scriptCacheId = UuidCreator.getRandomBased();
-        final File directory = CLASS_CACHE_LOCATION.resolve(UuidCreator.toString(scriptCacheId)).toFile();
+        final Path directory = CLASS_CACHE_LOCATION.resolve(UuidCreator.toString(scriptCacheId));
         createOrClearDirectory(directory);
         return directory;
     }
 
     public static void createScriptCache() {
-        final File classCacheDirectory = CLASS_CACHE_LOCATION.toFile();
-        createOrClearDirectory(classCacheDirectory);
+        createOrClearDirectory(CLASS_CACHE_LOCATION);
     }
 
-    private static void createOrClearDirectory(final File directory) {
-        if (directory.exists()) {
-            FileUtils.deleteRecursively(directory);
+    private static void createOrClearDirectory(final Path directory) {
+        if (Files.exists(directory)) {
+            FileUtils.deleteRecursively(directory.toFile());
         }
-        if (!directory.mkdirs()) {
+        try {
+            Files.createDirectories(directory);
+        } catch (IOException e) {
             throw new UncheckedDeephavenException(
-                    "Failed to create class cache directory " + directory.getAbsolutePath());
+                    "Failed to create class cache directory " + directory.toAbsolutePath(), e);
         }
     }
 
@@ -78,7 +80,7 @@
             final OperationInitializer operationInitializer,
             final ObjectTypeLookup objectTypeLookup,
             @Nullable final Listener changeListener) {
-        this(updateGraph, operationInitializer, objectTypeLookup, changeListener, newClassCacheLocation(),
+        this(updateGraph, operationInitializer, objectTypeLookup, changeListener, newClassCacheLocation().toFile(),
                 Thread.currentThread().getContextClassLoader());
     }
 
```

**Comment by rcaudy** [comment]
This feels like we're working around the fact that our processes expect exclusive ownership of their "cache" directory.

**Comment by devinrsmith** [comment]
I'm getting this too, which ultimately causes teardown NPEs.

```
Failed to create class cache directory /tmp/deephaven/cache/script-session-classes
io.deephaven.UncheckedDeephavenException: Failed to create class cache directory /tmp/deephaven/cache/script-session-classes
	at io.deephaven.engine.util.AbstractScriptSession.createOrClearDirectory(AbstractScriptSession.java:63)
	at io.deephaven.engine.util.AbstractScriptSession.createScriptCache(AbstractScriptSession.java:54)
	at io.deephaven.server.console.ScriptSessionCacheInit.<init>(ScriptSessionCacheInit.java:17)
	at io.deephaven.server.console.ConsoleModule.bindScriptSessionCacheInit(ConsoleModule.java:32)
	at io.deephaven.server.console.ConsoleModule_BindScriptSessionCacheInitFactory.bindScriptSessionCacheInit(ConsoleModule_BindScriptSessionCacheInitFactory.java:32)
	at io.deephaven.server.console.ConsoleModule_BindScriptSessionCacheInitFactory.get(ConsoleModule_BindScriptSessionCacheInitFactory.java:24)
	at io.deephaven.server.console.ConsoleModule_BindScriptSessionCacheInitFactory.get(ConsoleModule_BindScriptSessionCacheInitFactory.java:10)
	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
	at io.deephaven.server.console.NoConsoleSessionModule_BindNoLanguageSessionFactory.get(NoConsoleSessionModule_BindNoLanguageSessionFactory.java:46)
	at io.deephaven.server.console.NoConsoleSessionModule_BindNoLanguageSessionFactory.get(NoConsoleSessionModule_BindNoLanguageSessionFactory.java:14)
	at io.deephaven.server.console.NoConsoleSessionModule_BindScriptSessionFactory.get(NoConsoleSessionModule_BindScriptSessionFactory.java:37)
	at io.deephaven.server.console.NoConsoleSessionModule_BindScriptSessionFactory.get(NoConsoleSessionModule_BindScriptSessionFactory.java:13)
	at io.deephaven.server.session.ScriptSessionModule.provideScriptSession(ScriptSessionModule.java:31)
	at io.deephaven.server.session.ScriptSessionModule_ProvideScriptSessionFactory.provideScriptSession(ScriptSessionModule_ProvideScriptSessionFactory.java:44)
	at io.deephaven.server.session.ScriptSessionModule_ProvideScriptSessionFactory.get(ScriptSessionModule_ProvideScriptSessionFactory.java:34)
	at io.deephaven.server.session.ScriptSessionModule_ProvideScriptSessionFactory.get(ScriptSessionModule_ProvideScriptSessionFactory.java:13)
	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
	at io.deephaven.server.appmode.ApplicationInjector.loadApplicationFactory(ApplicationInjector.java:101)
	at io.deephaven.server.appmode.ApplicationInjector.run(ApplicationInjector.java:48)
	at io.deephaven.server.runner.DeephavenApiServer.startForUnitTests(DeephavenApiServer.java:242)
	at io.deephaven.server.runner.DeephavenApiServerTestBase.setUp(DeephavenApiServerTestBase.java:139)
	at io.deephaven.client.DeephavenSessionTestBase.setUp(DeephavenSessionTestBase.java:28)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.RunBefores.invokeMethod(RunBefores.java:33)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at io.grpc.testing.GrpcCleanupRule$1.evaluate(GrpcCleanupRule.java:153)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:112)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:40)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:60)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:52)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jdk.proxy1/jdk.proxy1.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$2.run(TestWorker.java:176)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:113)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```

I'm able to workaround it right now by setting `--max-workers 1`, but that's not a very satisfying solution.

**Review by rcaudy — APPROVED** [review]

**Pull request merged**